src/stream-tcp-reassemble.c

```diff
@@ -1838,14 +1838,22 @@ int StreamTcpReassembleHandleSegment(ThreadVars *tv, TcpReassemblyThreadCtx *ra_
         dir = UPDATE_DIR_BOTH;
     }
 
+    const bool reversed_before_ack_handling = p->flow != NULL && (p->flow->flags & FLOW_DIR_REVERSED) != 0; 
```

Comment:
Thanks again for this good bug report and investigation.

As this part of the code is treacherous, I wonder about the best fix.
Do we need the condition `p->flow != NULL` ?
Could we rather test `p->flowflags` change only ? as `PacketSwap` gets called also in `TCPProtoDetect`
Or have `StreamTcpReassembleAppLayer` somehow return if `AppLayerHandleTCPData` reversed the stream ?